File: Part1_강의자료/Ch08_투포인터/문제별코드/16472_고냥이/16472.py
def get_unique_alphabet_count(alphabet_count):
    return sum(1 for count in alphabet_count if count > 0)

N = int(input())
nyang = input()

alphabet_count = [0] * 26
next_idx = 0
max_length = 0

for i in range(len(nyang)):
    while next_idx < len(nyang):
        alphabet_count[ord(nyang[next_idx]) - ord('a')] += 1
        next_idx += 1
        if get_unique_alphabet_count(alphabet_count) > N:
            next_idx -= 1
            alphabet_count[ord(nyang[next_idx]) - ord('a')] -= 1
            break
    
    max_length = max(max_length, next_idx - i)
    alphabet_count[ord(nyang[i]) - ord('a')] -= 1

print(max_length)

File: Part1_강의자료/Ch08_투포인터/문제별코드/16472_고냥이/16472.cpp
#include <stdio.h>
#include <string.h>
#include <algorithm>
using namespace std;

int get_unique_alphabet_count(int *alphabet_count)
{
    int unique_alphabet_count = 0;
    for (int i = 0; i < 26; i++)
        if (alphabet_count[i] > 0)
            unique_alphabet_count++;
    return unique_alphabet_count;
}

int main()
{
    int N;
    char nyang[100005];

    scanf("%d", &N);
    scanf("%s", nyang);

    int alphabet_count[26];
    int next = 0;
    int max_length = 0;
    int len = strlen(nyang);
    for (int i = 0; i < len; i++)
    {
        while (next < len)
        {
            alphabet_count[nyang[next++] - 'a']++;
            if (get_unique_alphabet_count(alphabet_count) > N)
            {
                alphabet_count[nyang[--next] - 'a']--;
                break;
            }
        }
        max_length = max(max_length, next - i);
        alphabet_count[nyang[i] - 'a']--;
    }

    printf("%d\n", max_length);
}
